[bugfix] show results in query history

youngyjd opened the pull request:
This PR is to fix a bug caused by https://github.com/apache/incubator-superset/pull/5848

In my local environment, I have set up running query on main database as async. This is the demo for showing results.
![demo](https://user-images.githubusercontent.com/26216756/49396410-e4bbe000-f6ed-11e8-8d55-8c10a6f373bf.gif)

youngyjd committed:
[bugfix] show resules in query history

graceguo-supercat commented:
@youngyjd Thanks for the fix.

- see this line:
https://github.com/apache/incubator-superset/blob/69e8df404d46e35bf686cc92992d6e0415172d90/superset/assets/src/SqlLab/components/QueryAutoRefresh.jsx#L29

This line was added in #5848. Auto-refresh should run periodically as along as we have queries not finished in backend. For query in rendering state, it should be finished in backend and also fetched into browser already. Not sure we should still check status of these queries in the backend?

- see this line:
https://github.com/apache/incubator-superset/blob/fc3b68e23433522b3a6d7327cc95199bf01dcf96/superset/assets/src/SqlLab/constants.js#L12

This is all available status for the lifecycle of a query. If you add a state, can you also update this set? So that this new state will be taken care by other new or existed features.

- This is the lifecycle of a query (in my mind):
pending -> running -> success (or failed, stopped) ->fetching -> rendering
some state may too short to display:
 `success` is for a state that query is successfully finished in backend. 
 `fetching` is for a state that downloading data to browser.
I assume `rendering` means rendering data table. Can you make sure failed/stopped state also connected to rendering state? otherwise in error/stopped case, user will not see any updates.

youngyjd committed:
address comments

youngyjd commented:
@graceguo-supercat Thanks for reviewing. Updates:
1. Good catch. I have removed `rendering` state from QueryAutoRefresh, so after query results arrive in browser, it will not query backend again. This is really a good point.

2. https://github.com/apache/incubator-superset/blob/fc3b68e23433522b3a6d7327cc95199bf01dcf96/superset/assets/src/SqlLab/constants.js#L12 is used in [QuerySearch](https://github.com/apache/incubator-superset/blob/fc3b68e23433522b3a6d7327cc95199bf01dcf96/superset/assets/src/SqlLab/components/QuerySearch.jsx#L230). Not only `rendering` state is not in `STATUS_OPTION`, but also `stopped` and `fetching`. We will only need to add new state to `STATUS_OPTION` if we want to search for it in `QuerySearch`. I don't think we need to search for a query in `rendering` state as it is super fast that no query will remain in that state for a long time. Not adding `rendering` to `STATUS_OPTION` should be the same reason as why `fetching` is not in it either.

3. I digged into the sqllab code and the lifecycle I believe was designed as below:

- For `sync` queries: `running  -> success / failed / stopped`
- For `async` queries: `pending -> running -> fetching -> success` or `pending -> running -> fetching -> failed`

After I added the rendering state, it should look like:

- For `sync` queries: `running  -> rendering -> success / failed / stopped`
- For `async` queries: `pending -> running -> fetching -> rendering -> success` or `pending -> running -> fetching -> failed`

The `success` here I think means that query result is successfully rendered on UI, but not the real-time state for query running in backend. Backend query state is in `query.results` which may already says `success`, which browser side is showing `fetching` or `rendering`.

I am open to redesign the lifecycle of a query, but this should be in a new PR as this PR is to fix the bug without aiming to redesign a new lifecycle. What do you think?

youngyjd committed:
fix test

graceguo-supercat commented on superset/assets/src/SqlLab/components/ResultSet.jsx:
```diff
@@ -64,6 +64,14 @@ export default class ResultSet extends React.PureComponent {
       this.fetchResults(nextProps.query);
     }
   }
+  componentDidUpdate() {
```
I don't think this is the right place to update query state. this will cause whole component rerender.

youngyjd replied:
I tried, `success` state is unnecessary for sqllab to show query results. However, if you navigate to `Query History`, the state will show `rendering` instead of success. That's why I add this `componentDidUpdate()`.

graceguo-supercat commented on superset/assets/src/SqlLab/components/ResultSet.jsx:
```diff
@@ -163,7 +171,7 @@ export default class ResultSet extends React.PureComponent {
           </Alert>
         </div>);
-    } else if (query.state === 'success') {
+    } else if (['rendering', 'success'].indexOf(query.state) >= 0) {
```
And line 159, it should apply to `success` instead of `rendering`?

youngyjd replied:
yes. this part is called twice because of `componentDidUpdate()` in `ResultSet.jsx`. It is because if the state is `rendering`. I am not sure how we can change the `state` to `success` without `re-rendering`. What's your suggestion? This sound like falling into the same discussion in your comment below.

youngyjd commented:
@graceguo-supercat how about this? Let me try only have `rendering` state, but in https://github.com/apache/incubator-superset/blob/fc3b68e23433522b3a6d7327cc95199bf01dcf96/superset/assets/src/SqlLab/constants.js#L12 I change `success: 'success'` to `rendering: 'success'`, which means there won't be any state of `success`. All `rendering` state represents `success`. This may save us one round of component rerender.
